### PR TITLE
Simplify custom context usage & support custom autocomplete context

### PR DESCRIPTION
**Pull request by Luc1412**
## Summary

This pr simplifies custom context usage and adds custom autocomplete context support.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

**Comment by krittick** [comment]
@Luc1412 Do you have an example of how this would be used compared to the current functionality?

**Comment by Luc1412** [comment]
@krittick Currently, you need to overwrite the on_interaction function.

```py
async def on_interaction(self, interaction: discord.Interaction):
    if interaction.type not in (
            discord.InteractionType.application_command,
            discord.InteractionType.auto_complete
    ):
        return

    try:
        command = self.application_commands[interaction.data['id']]
    except KeyError:
        self.dispatch('unknown_command', interaction)
    else:
        if interaction.type is discord.InteractionType.auto_complete:
            return await command.invoke_autocomplete_callback(interaction)

        ctx = await self.get_application_context(interaction, cls=ApplicationContext)  # Inject the custom Context
        ctx.command = command
        self.dispatch('application_command', ctx)
        try:
            if await self.can_run(ctx, call_once=True):
                await ctx.command.invoke(ctx)
            else:
                raise discord.CheckFailure('The global check once functions failed.')
        except discord.DiscordException as exc:
            await ctx.command.dispatch_error(ctx, exc)
        else:
            self.dispatch('application_command_completion', ctx)
```
You are also not able to inject an AutocompleteContext.

With my fork, you can simply inject the custom Contexts like:
```py
async def on_interaction(self, interaction):
    await self.process_application_commands(interaction, application_context_cls=CustomContext, autocomplete_context_cls=ExampleContext)
```

**Comment by krittick** [comment]
> @krittick Currently, you need to overwrite the `on_interaction` and `process_application_commands` functions in order to inject your custom context class in `get_application_context`.
> 
> ```python
> async def on_interaction(self, interaction: discord.Interaction):
>     if interaction.type not in (
>             discord.InteractionType.application_command,
>             discord.InteractionType.auto_complete
>     ):
>         return
> 
>     try:
>         command = self.application_commands[interaction.data['id']]
>     except KeyError:
>         self.dispatch('unknown_command', interaction)
>     else:
>         if interaction.type is discord.InteractionType.auto_complete:
>             return await command.invoke_autocomplete_callback(interaction)
> 
>         ctx = await self.get_application_context(interaction, cls=ApplicationContext)  # Inject the custom Context
>         ctx.command = command
>         self.dispatch('application_command', ctx)
>         try:
>             if await self.can_run(ctx, call_once=True):
>                 await ctx.command.invoke(ctx)
>             else:
>                 raise discord.CheckFailure('The global check once functions failed.')
>         except discord.DiscordException as exc:
>             await ctx.command.dispatch_error(ctx, exc)
>         else:
>             self.dispatch('application_command_completion', ctx)
> ```
> 
> You are also not able to inject an AutocompleteContext.
> 
> With my fork, you can simply inject the custom Contexts like that:
> 
> ```python
> async def on_interaction(self, interaction):
>     await self.process_application_commands(interaction, application_context_cls=CustomContext, autocomplete_context_cls=ExampleContext)
> ```

This is super helpful to explain the context of the change, thank you!

**Comment by Luc1412** [comment]
It might make sense to make the custom classes keyword only. What do you think about it?

File: discord/bot.py [hunk 1]
```diff
@@ -367,6 +375,16 @@ async def process_application_commands(self, interaction: Interaction) -> None:
         -----------
         interaction: :class:`discord.Interaction`
             The interaction to process
+        application_context_cls
+            The factory class that will be used to create the context.
+            By default, this is :class:`.ApplicationContext`. Should a custom
+            class be provided, it must be similar enough to
+            :class:`.ApplicationContext`\'s interface.
```
Comment (BobDotCom) [inline discord/bot.py:382]:
```suggestion
            class be provided, it must be a subclass of :class:`.ApplicationContext`.
```

File: discord/bot.py [hunk 2]
```diff
@@ -367,6 +375,16 @@ async def process_application_commands(self, interaction: Interaction) -> None:
...
+        autocomplete_context_cls
+            The factory class that will be used to create the context.
+            By default, this is :class:`.AutocompleteContext`. Should a custom
+            class be provided, it must be similar enough to
+            :class:`.AutocompleteContext`\'s interface.
```
Comment (BobDotCom) [inline discord/bot.py:387]:
```suggestion
            class be provided, it must be a subclass of :class:`.AutocompleteContext`.
```

File: discord/commands/commands.py [hunk 3]
```diff
@@ -527,17 +527,24 @@ async def _invoke(self, ctx: ApplicationContext) -> None:
...
 
-    async def invoke_autocomplete_callback(self, interaction: Interaction):
-        values = { i.name: i.default for i in self.options }
+    async def invoke_autocomplete_callback(
+            self,
+            interaction: Interaction,
+            cls: Optional[Type[AutocompleteContext]] = None
+    ):
+        if cls is None:
+            cls = AutocompleteContext
```
Comment (BobDotCom) [inline discord/commands/commands.py:536]:
```suggestion
            cls: AutocompleteContext = AutocompleteContext
    ):
```

**Review by BobDotCom — CHANGES_REQUESTED** [review]
Some fixes to make this more usable (untested).

Regardless, I think we should update the implementation of this by introducing the bot.get_command_context and get_autocomplete_context methods.

File: discord/commands/commands.py (hunk 3 above)
Comment (Luc1412) [reply discord/commands/commands.py:536]: Think that's wrong. Should be `Type[AutocompleteContext]`

**Comment by CodeWithSwastik** [comment]
You don't need to overwrite `on_interaction` xD

```py
    async def get_application_context(self, interaction, cls=None):
        return await super().get_application_context(interaction, cls=cls or CustomContext)
```

This is how you are supposed to do it. I'll implement a `get_application_autocomplete_context` method. I think this should be closed.

**Comment by Luc1412** [comment]
Uhm that's true. That can simplify things. Might be also worth to adding an example or hint in the docs how to use custom context

**Comment by CodeWithSwastik** [comment]
I think that already exists https://github.com/Pycord-Development/pycord/blob/master/examples/custom_context.py

**Comment by CodeWithSwastik** [comment]
Alright I'll close this for now

**Pull request closed without merging**